scoped-mutex/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+/// just implements the mutex mechanism. For most uses you should use [`BlockingMutex`]
+/// instead, which is generic over a RawMutex and contains the protected data.
+///
+/// # Safety
+///
+/// RawMutex implementations must ensure that, while locked, no other thread can lock
+/// the RawMutex concurrently. This can usually be implemented using an [`AtomicBool`]
+/// to track the "taken" state. See [crate::impls] for examples of correct implementations.
+///
+/// [`AtomicBool`]: core::sync::atomic::AtomicBool
+///
+/// Unsafe code is allowed to rely on this fact, so incorrect implementations will cause undefined behavior.
+///
+/// # Implementation Note:
+///
+/// This is actually a marker trait for types that implement [`ScopedRawMutex`] and
+/// [`ConstInit`]. This is to allow cases where a mutex cannot be created in const
+/// context, for example some runtime/OS mutexes, as well as testing mutexes like
+/// those from `loom`.
+///
+/// If you are implementing your own RawMutex primitive, you should implement the
+/// [`ScopedRawMutex`] and [`ConstInit`] traits, and rely on the blanket impl
+/// of `impl<T: ScopedRawMutex + ConstInit> RawMutex for T {}`.
+pub trait ConstScopedRawMutex: ScopedRawMutex + ConstInit {}
+
+impl<T: ScopedRawMutex + ConstInit> ConstScopedRawMutex for T {}
+
+pub trait ConstInit {
+    /// Create a new instance.
+    ///
+    /// This is a const instead of a method to allow creating instances in const context.
+    const INIT: Self;
+}
+
+/// Raw mutex trait.
+///
+/// This mutex is "raw", which means it does not actually contain the protected data, it
+/// just implements the mutex mechanism. For most uses you should use [`BlockingMutex`]
+/// instead, which is generic over a ScopedRawMutex and contains the protected data.
+///
+/// # Safety
+///
+/// ScopedRawMutex implementations must ensure that, while locked, no other thread can lock
+/// the RawMutex concurrently. This can usually be implemented using an [`AtomicBool`]
+/// to track the "taken" state. See [crate::impls] for examples of correct implementations.
+///
+/// Unsafe code is allowed to rely on this fact, so incorrect implementations will cause undefined behavior.
+///
+/// [`AtomicBool`]: core::sync::atomic::AtomicBool
+pub unsafe trait ScopedRawMutex {
+    /// Lock this `ScopedRawMutex`, calling `f()` after the lock has been acquired, and releasing
+    /// the lock after the completion of `f()`.
+    ///
+    /// If this was successful, `Some(R)` will be returned. If the mutex was already locked,
+    /// `None` will be returned
+    #[must_use]
+    fn try_lock<R>(&self, f: impl FnOnce() -> R) -> Option<R>;
+
+    /// Lock this `ScopedRawMutex`, calling `f()` after the lock has been acquired, and releasing
+    /// the lock after the completion of `f()`.
+    ///
+    /// Panics if the lock is already locked.
+    fn lock<R>(&self, f: impl FnOnce() -> R) -> R {
+        self.try_lock(f).expect("Attempted to take lock re-entrantly")
+    }
```

Comment:
I don't think this is correct. This method should _block_ if the lock is already locked, instead.  What you've got here is a `RefCell`, not a blocking mutex.

```suggestion
    /// Lock this `ScopedRawMutex`, calling `f()` after the lock has been acquired, and releasing
    /// the lock after the completion of `f()`.
    ///
    /// If the lock is held by another thread or CPU core, this method will block
    /// until the lock is released, and then acquire the lock. The mechanism used
    /// to block depends on the implementation of this trait.
    ///
    /// To attempt to lock the mutex *without* blocking, use
    /// [`ScopedRawMutex::try_lock`].
    fn lock<R>(&self, f: impl FnOnce() -> R) -> R;
```

I also don't really think we should have a default impl for this or not, since it's the _way_ we block while waiting for the lock to become available that's the main thing implementations will be providing. Checking if the lock is locked or not is basically always going to be the same except in the totally-no-op case. If we were going to provide one, though, it should probably look more like:
```rust 
fn lock<R>(&self, f: impl FnOnce() -> R) -> R{
    // work around the fact that `f` is a `FnOnce`; we're only ever going to
    // call it once, but we need to pass it into `try_lock` multiple times regardless.
    let mut f = Some(f);
    loop {
       if let Some(r) = self.try_lock(|| (f.take().unwrap())()) {
           return r;
       }
       core::hint::spin_loop();
    }
}
```
but, this isn't even the most efficient implementation of a spinlock, which would instead probably do something more like [this](https://github.com/hawkw/mycelium/blob/c4992526f87f5c38813e6671baf99089fa24d7f0/maitake-sync/src/spin/mutex.rs#L106-L114), using `Relaxed` loads while waiting for the lock to be unlocked, and backing off in the spin loop.

So, i'm not sure if it's a good idea to provide a default implementation here.